Mathlib/RingTheory/Congruence/Hom.lean

```diff
@@ -0,0 +1,476 @@
+theorem lift_comp_mk' (H : c ≤ ker f) : (c.lift f H).comp c.mk' = f := rfl
+
+/-- Given a homomorphism `f` from the quotient of a ring by a ring congruence
+relation, `f` equals the homomorphism on the quotient induced by `f` composed
+with the natural map from the ring to the quotient. -/
+theorem lift_apply_mk' (f : c.Quotient →+* P) :
+    (c.lift (f.comp c.mk') fun x y h => show f ↑x = f ↑y by rw [c.eq.2 h]) = f := by
+  ext x; rcases x with ⟨⟩; rfl
+
+/-- Homomorphisms on the quotient of a ring by a ring congruence relation are
+equal if they are equal on elements that are coercions from the ring. -/
+@[ext]
+theorem Quotient.hom_ext {f g : c.Quotient →+* P} (h : f.comp c.mk' = g.comp c.mk') : f = g :=
+  DFunLike.ext _ _ <| c.mk'_surjective.forall.mpr fun x ↦ by exact congr($h x)
+
+/-- The uniqueness part of the universal property for quotients of rings. -/
+theorem lift_unique (H : c ≤ ker f) (g : c.Quotient →+* P) (Hg : g.comp c.mk' = f) :
+    g = c.lift f H :=
+  Quotient.hom_ext (by aesop)
+
+/-- Surjective ring homomorphisms constant on a the equivalence classes
+of a ring congruence relation induce a surjective homomorphism on the quotient. -/
+theorem lift_surjective_of_surjective (h : c ≤ ker f) (hf : Surjective f) :
+    Surjective (c.lift f h) :=
+  AddCon.lift_surjective_of_surjective _ hf
+
+variable (c f)
+
+/-- Given a ring homomorphism `f` from `M` to `P`, the kernel of `f` is the
+unique ring congruence relation on `M` whose induced map from the quotient of
+`M` to `P` is injective. -/
+theorem ker_eq_lift_of_injective (H : c ≤ ker f) (h : Injective (c.lift f H)) : ker f = c :=
+  ext'' <| Setoid.ker_eq_lift_of_injective f H h
+
+variable {c}
+
+/-- The homomorphism induced on the quotient of a ring by the kernel of a ring homomorphism. -/
+def kerLift : (ker f).Quotient →+* P :=
+  (ker f).lift f fun _ _ => id
+
+variable {f}
+
+/-- The diagram described by the universal property for quotients of rings, when
+the ring congruence relation is the kernel of the homomorphism, commutes. -/
+theorem kerLift_mk (x : M) : kerLift f x = f x :=
+  rfl
+
+/-- A ring homomorphism `f` induces an injective homomorphism on the quotient by `f`'s kernel. -/
+theorem kerLift_injective (f : M →+* P) : Injective (kerLift f) :=
+  AddCon.kerLift_injective (f : M →+ P)
+
+/-- Given ring congruence relations `c, d` on a ring such that `d` contains `c`,
+`d`'s quotient map induces a homomorphism from the quotient by `c` to the
+quotient by `d`. -/
+def map (c d : RingCon M) (h : c ≤ d) : c.Quotient →+* d.Quotient :=
+  (c.lift d.mk') fun x y hc => show (ker d.mk') x y from (ker_mk'_eq d).symm ▸ h hc
+
+/-- Given ring congruence relations `c, d` on a ring such that `d` contains `c`,
+the definition of the homomorphism from the quotient by `c` to the quotient by
+`d` induced by `d`'s quotient map. -/
+theorem map_apply {c d : RingCon M} (h : c ≤ d) (x) :
+    c.map d h x = c.lift d.mk' (fun _ _ hc => d.eq.2 <| h hc) x :=
+  rfl
+
+end
+
+section
```

Comment:
Note to self: I have reviewed until here.

This PR is slightly on the large side and it might be useful to split it up. (perhaps here?)